distributed/worker.py

```diff
@@ -1555,10 +1586,104 @@ def transition(self, ts, finish, **kwargs):
             self.validate_task(ts)
         self._notify_plugins("transition", ts.key, start, state or finish, **kwargs)
 
-    def transition_waiting_flight(self, ts, worker=None):
+    def transition_new_waiting(self, ts):
         try:
             if self.validate:
-                assert ts.state != "flight"
+                assert ts.state == "new"
+                assert ts.runspec is not None
+                assert not ts.who_has
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_new_fetch(self, ts):
+        try:
+            if self.validate:
+                assert ts.state == "new"
+                assert ts.runspec is None
+
+            for dependent in ts.dependents:
+                dependent.waiting_for_data.add(ts.key)
+
+            self.data_needed.append(ts.key)
+            self.waiting_for_data_count += 1
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_fetch_waiting(self, ts, runspec=None):
```

Comment:
`runspec` is not actually allowed to be `None` here, is it?

Reply:
No, it isn't, I can add a check for that

Reply:
My point is rather why the default value if it must be provided anyhow

Reply:
Ah, gotcha.  I was just following the convention in the other transition functions, but you're right.  Much better to be explicit here that it is a required value.